components/tidb_query/src/batch/executors/table_scan_executor.rs

```diff
@@ -149,32 +152,40 @@ impl ScanExecutorImpl for TableScanExecutorImpl {
         let columns_len = self.schema.len();
         let mut columns = Vec::with_capacity(columns_len);
 
-        if let Some(handle_index) = self.handle_index {
-            // PK is specified in schema. PK column should be decoded and the rest is in raw format
-            // like this:
-            // non-pk non-pk non-pk pk non-pk non-pk non-pk
-            //                      ^handle_index = 3
-            //                                             ^columns_len = 7
-
-            // Columns before `handle_index` (if any) should be raw.
-            for _ in 0..handle_index {
+        // If there are any PK columns, for each of them, fill non-PK columns before it and push the
```

Comment:
How about using the following code instead of L155-191
```
        let mut last_handle_index = 0usize;
        for id in 0..columns_len {
            let is_handle = self
                .handle_indices
                .get(last_handle_index)
                .map_or(false, |handle_idx| *handle_idx == id);
            if is_handle {
                last_handle_index += 1;
                columns.push(LazyBatchColumn::decoded_with_capacity_and_tp(
                    scan_rows,
                    EvalType::Int,
                ));
            } else {
                columns.push(LazyBatchColumn::raw_with_capacity(scan_rows));
            }
        }
```

Reply:
This is the same as my first idea. It needs an indexing to the vector and a `if` for each column. You can see that the old code is trying to avoid it, although I don't think it matters much. @breeswish what's your idea?

Reply:
The old code is branch-less in each iteration.

Reply:
@breeswish Yes that's what I considered in my new code. My question is, do you prefer @AndreMouche 's simlpler, more readable code with branches in each iteration?

Reply:
I don't have strong opinions 🤔

Reply:
@AndreMouche Let's keep the current code? since the PR has been approved.